SIS_import/sis_upload.rb

```diff
@@ -154,69 +249,92 @@ def parseJson(data)
-	end
-end
+	upload_error = "Cannot find logs directory " + outputDirectory
 
-if (!uploadError)
-	## if there is no upload error
-	# move file to archive directory after processing
-	FileUtils.mv(fileName, archiveDirectory+currentFileBaseName)
-	p "SIS upload finished with " + fileName
-	exit
-else
 	## check first about the environment variable setting for MAILTO '
-	p " use the environment variable for sending out error messages #{ENV['MAILTO']}"
+	p "Use the environment variable 'MAILTO' for sending out error messages to #{ENV['MAILTO']}"
 	## send email to support team with the error message
-	`echo #{uploadError} | mail -s "#{server} Upload Error" #{ENV['MAILTO']}`
-	abort(uploadError)
+	`echo #{upload_error} | mail -s "#{server} Upload Error" #{ENV['MAILTO']}`
+	abort(upload_error)
+else
+	#open the output file
+	begin
+		# get output file name
+		output_file_base_name = "Canvas_upload_" + Time.new.strftime("%Y%m%d%H%M%S")
+		outputFile = File.open(outputDirectory + output_file_base_name + ".txt", "w")
+
+		if (Dir[currentDirectory].length != 1)
+			## working directory
+			upload_error = "Cannot find current working directory " + currentDirectory + "."
+		elsif (Dir[archiveDirectory].length != 1)
+			## archive directory
+			upload_error = "Cannot find archive directory " + archiveDirectory + "."
+		elsif (Dir[tokenFile].length != 1)
+			## token file
+			upload_error = "Cannot find token file #{tokenFile}."
+		else
+			File.open(tokenFile, 'r') do |tFile|
+				while line = tFile.gets
+					# only read the first line, which is the token value
+					token=line.strip
+					break
+				end
+			end
+			if (token=="")
+				upload_error="Empty token for Canvas upload."
+			end
+		end
+
+		if (!upload_error)
+			# the canvas import zip file
+			fileNames = Dir[currentDirectory+ "Canvas_Extract_*.zip"];
+			if (fileNames.length == 0)
+				## cannot find zip file to upload
+				upload_error = "Cannot find SIS zip file."
+			elsif (fileNames.length > 1)
+				## there are more than one zip file
+				upload_error = "There are more than one SIS zip files to be uploaded."
+			elsif
+				## get the name of file to process
+				fileName=fileNames[0]
+				currentFileBaseName = File.basename(fileName, ".zip")
+
+				## checksum verification step
+				upload_error = verify_checksum(currentDirectory + currentFileBaseName)
+
+				if (!upload_error)
+					# upload the file to canvas server
+					upload_error = upload_to_canvas(currentDirectory, fileName, token, server, outputDirectory, outputFile, output_file_base_name)
+				end
+			end
+		end
+
+		if (upload_error)
+			# write upload error string into file
+			outputFile.write(upload_error)
+
+			## check first about the environment variable setting for MAILTO '
+			p "Use the environment variable 'MAILTO' for sending out error messages to #{ENV['MAILTO']}"
+			## send email to support team with the error message
+			`echo #{upload_error} | mail -s "#{server} Upload Error" #{ENV['MAILTO']}`
```

Comment:
Just as Ruby offers HTTP clients, it also offers SMTP clients.  Replace this external call to mail with a function call to a SMTP function/method.